domain/runaway.go

```diff
@@ -0,0 +1,525 @@
+			return
+		}
+		leftRows = make([][]types.Datum, len(rows))
+		for i, row := range rows {
+			leftRows[i] = row.GetDatumRow(tb.KeyColumnTypes)
+		}
+
+		for len(leftRows) > 0 {
+			var delBatch [][]types.Datum
+			if len(leftRows) < runawayRecordGCBatchSize {
+				delBatch = leftRows
+				leftRows = nil
+			} else {
+				delBatch = leftRows[0:runawayRecordGCBatchSize]
+				leftRows = leftRows[runawayRecordGCBatchSize:]
+			}
+			sql, err := sqlbuilder.BuildDeleteSQL(tb, delBatch, expiredTime)
+			if err != nil {
+				logutil.BgLogger().Error(
+					"build delete SQL failed when deleting system table",
+					zap.Error(err),
+					zap.String("table", tb.Schema.O+"."+tb.Name.O),
+				)
+				return
+			}
+
+			_, err = do.execRestrictedSQL(sql, nil)
+			if err != nil {
+				logutil.BgLogger().Error(
+					"delete SQL failed when deleting system table", zap.Error(err), zap.String("SQL", sql),
+				)
+			}
+		}
+	}
+}
+
+func (do *Domain) runawayWatchSyncLoop() {
+	defer util.Recover(metrics.LabelDomain, "runawayWatchSyncLoop", nil, false)
+	runawayWatchSyncTicker := time.NewTicker(runawayWatchSyncInterval)
+	updateFn := func() error {
+		records, err := do.runawaySyncer.getNewWatchRecords()
+		if err != nil {
+			logutil.BgLogger().Error("try to get new runaway watch", zap.Error(err))
+			return err
+		}
+		for _, r := range records {
+			do.runawayManager.AddWatch(r)
+		}
+		doneRecords, err := do.runawaySyncer.getNewWatchDoneRecords()
+		if err != nil {
+			logutil.BgLogger().Error("try to get done runaway watch", zap.Error(err))
+			return err
+		}
+		for _, r := range doneRecords {
+			do.runawayManager.RemoveWatch(r)
+		}
+		return nil
+	}
+	for {
+		select {
+		case <-do.exit:
+			return
+		case <-do.runawaySyncer.notifyChan:
+			err := updateFn()
+			do.runawaySyncer.doneChan <- err
+		case <-runawayWatchSyncTicker.C:
+			updateFn()
+		}
+	}
+}
+
+func (do *Domain) AddRunawayWatch(record *resourcegroup.QuarantineRecord) (int64, error) {
+	if err := do.handleRunawayWatch(record); err != nil {
+		return 0, err
+	}
+	timer := time.NewTimer(10 * time.Second)
+	ticker := time.NewTicker(time.Second)
+	defer func() {
+		ticker.Stop()
+		timer.Stop()
+	}()
+	for {
+		select {
+		case <-timer.C:
+			return 0, errors.Errorf("add runaway watch timeout")
+		case <-ticker.C:
+			r := do.runawayManager.GetWatchInWatchList(record.GetRecordKey())
+			if r.ID > 0 {
+				return r.ID, nil
+			}
+		case err := <-do.runawaySyncer.doneChan:
+			if err != nil {
+				continue
+			}
+			r := do.runawayManager.GetWatchInWatchList(record.GetRecordKey())
+			if r.ID > 0 {
+				return r.ID, nil
+			}
+		}
+	}
+}
+
+func (do *Domain) RemoveRunawayWatch(recordID int64) error {
+	records, err := do.runawaySyncer.getWatchRecordByID(recordID)
+	if err != nil {
+		return err
+	}
+	if len(records) != 1 {
+		return errors.Errorf("no runaway watch with the specific ID")
+	}
+	err = do.handleRunawayWatchDone(records[0])
```

Comment:
The sql in `handleRunawayWatchDone` and `handleRemoveStaleRunawayWatch` should be executed in the same txn to ensure atomic

Reply:
fix